runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/graph/GreedyPipelineFuserTest.java

```diff
@@ -0,0 +1,857 @@
+                        FunctionSpec.newBuilder()
+                            .setUrn(PTransformTranslation.PAR_DO_TRANSFORM_URN)
+                            .setPayload(
+                                ParDoPayload.newBuilder()
+                                    .setDoFn(SdkFunctionSpec.newBuilder().setEnvironmentId("py"))
+                                    .build()
+                                    .toByteString()))
+                    .build())
+            .putPcollections(
+                "pyParDo.out", PCollection.newBuilder().setUniqueName("pyParDo.out").build())
+            .putTransforms(
+                "goParDo",
+                PTransform.newBuilder()
+                    .putInputs("input", "flatten.out")
+                    .putOutputs("output", "goParDo.out")
+                    .setSpec(
+                        FunctionSpec.newBuilder()
+                            .setUrn(PTransformTranslation.PAR_DO_TRANSFORM_URN)
+                            .setPayload(
+                                ParDoPayload.newBuilder()
+                                    .setDoFn(SdkFunctionSpec.newBuilder().setEnvironmentId("go"))
+                                    .build()
+                                    .toByteString()))
+                    .build())
+            .putPcollections(
+                "goParDo.out", PCollection.newBuilder().setUniqueName("goParDo.out").build())
+            .putEnvironments("go", Environment.newBuilder().setUrl("go").build())
+            .putEnvironments("py", Environment.newBuilder().setUrl("py").build())
+            .build();
+    FusedPipeline fused =
+        GreedyPipelineFuser.fuse(Pipeline.newBuilder().setComponents(components).build());
+
+    assertThat(
+        fused.getRunnerExecutedTransforms(),
+        containsInAnyOrder(
+            PipelineNode.pTransform("pyImpulse", components.getTransformsOrThrow("pyImpulse")),
+            PipelineNode.pTransform("goImpulse", components.getTransformsOrThrow("goImpulse"))));
+    assertThat(
+        fused.getFusedStages(),
+        containsInAnyOrder(
+            ExecutableStageMatcher.withInput("goImpulse.out")
+                .withOutputs("flatten.out")
+                .withTransforms("goRead", "flatten"),
+            ExecutableStageMatcher.withInput("pyImpulse.out")
+                .withOutputs("flatten.out")
+                .withTransforms("pyRead", "flatten"),
+            ExecutableStageMatcher.withInput("flatten.out")
+                .withNoOutputs()
+                .withTransforms("goParDo"),
+            ExecutableStageMatcher.withInput("flatten.out")
+                .withNoOutputs()
+                .withTransforms("pyParDo")));
+  }
+
+  /*
+   * impulseA -> .out -> goRead -> .out \
+   *                                     -> flatten -> .out -> goParDo -> .out
+   * impulseB -> .out -> pyRead -> .out /
+   *
+   * becomes
+   * (impulseA.out) -> goRead -> goRead.out -> flatten -> flatten.out -> goParDo
```

Comment:
Looking at this example, we are going to have to add anything with state/timers to be like side inputs where it is in its own fused stage.

Reply:
ACK.